### ISSUE TITLE
Compilation error on CollectionMapping. Wrong capture generated

### ISSUE THREAD

**Issue by soberich** [mapstruct/mapstruct#1453]
Hi,
```
protected Collection<? extends PaymentDTO> paymentListToPaymentDTOCollection(List<Payment> list) {
        if ( list == null ) {
            return null;
        }
        // Capture cause compilation error, should be "Collection<PaymentDTO>"
        Collection<? extends PaymentDTO> collection = new ArrayList<PaymentDTO>( list.size() );
        for ( Payment payment : list ) {
            collection.add( toDto( payment ) ); // <== here
        }

        return collection;
    }

```

error
```
error: method add in interface Collection<E> cannot be applied to given types;
            collection.add( toDto( payment ) );
                      ^
  required: CAP#1
  found: PaymentDTO
  reason: argument mismatch; PaymentDTO cannot be converted to CAP#1
  where E is a type-variable:
    E extends Object declared in interface Collection
  where CAP#1 is a fresh type-variable:
    CAP#1 extends PaymentDTO from capture of ? extends PaymentDTO
1 error
```

toDto mapping method is just
```
    @Mapping(target = "term",                        ignore = true)
    PaymentDTO toDto(Payment payment);
```
which generates
```
    @Override
    public PaymentDTO toDto(Payment payment) {
        if ( payment == null ) {
            return null;
        }

        PaymentDTOBuilder paymentDTO = PaymentDTO.builder();

        paymentDTO.type( payment.getType() ); // I am using getters with getX for Intellij to advise, since plugin can't see fluent getters , later I just uncomment @Accessors(fluent=true)
        paymentDTO.amount( payment.getAmount() );
        paymentDTO.description( payment.getDescription() );

        return paymentDTO.build();
    }
```

The Solution for now is adder method and ADDER_PREFFERED which works fine.

**Comment by soberich** [comment]
/*
     * Error
     * */

target is `List<@NotNull @Valid Payment>` source is as well `List<PaymentDTO>`
```
    @Mapping(target = "payments",                    source = "auctionBean.payments")
    @Mapping(target = "limits",                      source = "auctionBean.limits")
    AuctionDTO toDto( AuctionBean auctionBean, AuctionParam auctionParam);

    @Override
    public AuctionDTO toDto(AuctionBean auctionBean, AuctionParam auctionParam) {
        if ( auctionBean == null && auctionParam == null ) {
            return null;
        }

        AuctionDTOBuilder auctionDTO = AuctionDTO.builder();

        if ( auctionBean != null ) {
            auctionDTO.totalProjectDuration( auctionBean.totalProjectDuration() );
            auctionDTO.jurisdiction( auctionBean.jurisdiction() );
            //HERE
            auctionDTO.payments( paymentListToPaymentDTOCollection( auctionBean.payments() ) );
            //HERE
            String insuranceType = auctionBeanGeneralDeterminantsInsuranceType( auctionBean );
            if ( insuranceType != null ) {
                auctionDTO.insuranceType( insuranceType );
            }
            ... more stuff
        }

        return auctionDTO.build();
    }
```
The below method with error is called ONLY from method above
```
    protected Collection<? extends PaymentDTO> paymentListToPaymentDTOCollection(List<Payment> list) {
        if ( list == null ) {
            return null;
        }
        // WRONG CAPTURE
        Collection<? extends PaymentDTO> collection = new ArrayList<PaymentDTO>( list.size() );
        for ( Payment payment : list ) {
            collection.add( toDto( payment ) );
        }                  //^
                          //below the toDto method just in case

        return collection;
    }
```
Here is toDto(..) just in case
```
    @Override
    public PaymentDTO toDto(Payment payment) {
        if ( payment == null ) {
            return null;
        }

        PaymentDTOBuilder paymentDTO = PaymentDTO.builder();

        paymentDTO.type( payment.type() );
        paymentDTO.amount( payment.amount() );
        paymentDTO.description( payment.description() );

        return paymentDTO.build();
    }
```

**Comment by filiphr** [comment]
How does the setter look like for the payments in the `ActionDTOBuilder`. I think we would be able to reproduce it if I know that. Thanks a lot for all your inputs until now

**Comment by soberich** [comment]
Oh yeah. 
here we go. 
This is the one
```
        public AuctionDTO.AuctionDTOBuilder payments(Collection<? extends PaymentDTO> payments) {
            if (this.payments == null) this.payments = new ArrayList<PaymentDTO>();
            this.payments.addAll(payments);
            return this;
        }
```
This is what Lombok creates, so clearly Mapstruct inherits this signature.

**Comment by filiphr** [comment]
That's what I suspected. Thanks a lot for the details, it will help us write a reproducing test. It won't be too difficult to fix this.

Btw, this is not linked to the builders support. The same would happen with a normal setter as well

**Comment by filiphr** [comment]
@soberich you said:

>The Solution for now is adder method and ADDER_PREFFERED which works fine.

I am curious does your adder not return the builder type? I am asking because I noticed #1449 for builders actually

**Comment by soberich** [comment]
I've seen your last commit on that and simply copied that line to my CustomStrategy where you escape adder when resolving setters.

**Comment by filiphr** [comment]
>I've seen your last commit on that and simply copied that line to my CustomStrategy where you escape adder when resolving setters.

OK for the moment I thought I was tripping 😄. 

In any case that line is wrong as `address(String address)` would not be considered as an adder. We need to figure out a better solution

**Comment by filiphr** [comment]
@soberich I think I have solved the issue. Can you please try out the PR #1465?

**Comment by soberich** [comment]
I will try very soon. 
Meanwhile I have new, please see https://github.com/mapstruct/mapstruct/issues/1467

**Comment by filiphr** [comment]
@soberich this has been fixed in PR #1465 it should be in the snapshot repository in 30 minutes max
